Nightly allinone

FiftyTifty opened the pull request:
This branch is based on Nightly, with my additions to ShaderManager.cpp, ShaderManager.h, and the .toml file. I got permission to port his Oblivion ENB preset here: https://forums.nexusmods.com/index.php?/topic/5083940-photorealistic-tamriel-beta-0-1-0/page-73#entry127083570

It's all contained within one shader in the menu called AllInOne, with the individual effects separated to individual .hlsl files which are included.

FiftyTifty committed:
Added port of L00Ping's NAO Post Processing

No removals to any of the code, just additions.

mchaptel reviewed (changes requested):
Looking really good, there's a couple of things I think we could fix before merging though!

First, I think that looks like a lot of includes and not all of it seems used, maybe there's a way to make it more self contained? By only adding the functions that are actually used? And maybe we could add the functions directly in the main shader, as includes are mostly good for sharing code between effects which won't happen here probably.

Second, it seems like toml libs got added to the history, it would be good if we could remove that so as to not mess up the repo too much.

FiftyTifty commented:
For the includes, IIRC they're all used. I separated out the HLSL code into separate .hlsl files to make it easier to work on them, as well as keeping the code self contained and compartmentalized. Can be just ctrl+c ctrl+v'd into one big file, but that will be messy and awful to read.

Yeah I don't know how the .toml stuff got added, I had this issue before when making edits to the code. What's the way to get rid of them?

FiftyTifty commented:
Ahh, I think I get what you mean now. Yeah I forgot to remove the LUT helper, that's a WiP I plan to continue work on further down the line. And there's probably stuff in aioNAL_utilities I can remove. Anything else you noticed?

FiftyTifty committed:
Commented out unused utilities

Commented out a few functions in aioNAL_Utilities, as well as the entire copy-paste of aioLUT (haven't ported it from ENB yet)

FiftyTifty commented:
@mchaptel That's the shader code taken care of. Don't quite know how to approach the .toml stuff though

FiftyTifty committed:
Updated gitignore to exclude TESReloaded/lib

FiftyTifty committed:
Added wildcard to ignore libraries

FiftyTifty committed:
Fixed wrong interior variable names

FiftyTifty committed:
Split the exterior / interior lerps

mchaptel commented:
> Ahh, I think I get what you mean now. Yeah I forgot to remove the LUT helper, that's a WiP I plan to continue work on further down the line. And there's probably stuff in aioNAL_utilities I can remove. Anything else you noticed?

I would really prefer one single shader file for this (YAGNI principle ;) ) it clogs up the repo for one effect and I think we have a redundant luma formula in there etc. I would prefer to not multiply includes unless reused across many effects.
I'll check the latest changes!

about the toml stuff, not too sure, maybe we could cherry pick the relevant commits to a clean branch just for merging?

Another thing, would you consider having Night/Day subcategories in order to keep the organisation of other shaders? 
Instead of having a Night and Day value for each setting in the same?

FiftyTifty commented:
> I would really prefer one single shader file for this (YAGNI principle ;) ) it clogs up the repo for one effect and I think we have a redundant luma formula in there etc. I would prefer to not multiply includes unless reused across many effects. I'll check the latest changes!
> 
> about the toml stuff, not too sure, maybe we could cherry pick the relevant commits to a clean branch just for merging?

Aye fair. I'll just have to suck it up and merge them together. Is it possible to just toss them into their own functions (in the same .hlsl) to keep things tidy?

The thing is, whenever I open the SLN file after `git clone`, the toml files are always changed by VS for some reason. Strange that Github doesn't allow us to just pick and choose the changed files to put into the pull...

FiftyTifty commented:
>Another thing, would you consider having Night/Day subcategories in order to keep the organisation of other shaders?
Instead of having a Night and Day value for each setting in the same?

It's a personal preference really, I like having all the settings for an effect together, as it involves less screwing around with arrow keys in the menu. I'd rather not, but ofc if there are standards, then the changes gotta be made.

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/Includes/aioNAL_Utilities.hlsl:
```diff
@@ -0,0 +1,84 @@
+      hue = 0.0;
+   else
+   {         
+      float deltaR = (((fmax - color.r) / 6.0) + (delta / 2.0)) / delta;
```
it would be better to do 1 calculation with a float3

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/Includes/aioNAL_Utilities.hlsl:
```diff
@@ -0,0 +1,84 @@
+      float deltaG = (((fmax - color.g) / 6.0) + (delta / 2.0)) / delta;
+      float deltaB = (((fmax - color.b) / 6.0) + (delta / 2.0)) / delta;
+
+      if (color.r == fmax )
```
maybe here we could use a couple of ternaries or lerps? maybe even a float3 to do all 3 calculations at once, then select the channel with a ternary

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/Includes/aioNAL_Utilities.hlsl:
```diff
@@ -0,0 +1,84 @@
+         hue = (2.0 / 3.0) + deltaG - deltaR;
+   }
+      
+   if (hue < 0.0)
```
bit confused about this. Is it trying to saturate?

FiftyTifty replied:
I don't know the intricacies of the code, unfortunately. I know kind of what the variables do, but that's as far as I go with HLSL. That function is something to do with converting from RGB to HSL, but in a different way to change the colour?

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/Includes/aioNAL_Utilities.hlsl:
```diff
@@ -0,0 +1,84 @@
+   return dot( gv, float3(0.2125, 0.7154, 0.0721) );
+}
+
+float smootherstep(float edge0, float edge1, float x)
```
maybe we could define this in Includes/Helpers.hlsl

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/Includes/aioNAL_Utilities.hlsl:
```diff
@@ -0,0 +1,84 @@
+		
+	float grayValue(float3 gv)
+{
+   return dot( gv, float3(0.2125, 0.7154, 0.0721) );
```
I think this already exists in Includes/Helpers

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/Includes/aioNAL_Utilities.hlsl:
```diff
@@ -0,0 +1,84 @@
+/*
```
we can remove commented code from the commits

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/Includes/aioNAL_ColorFilter.hlsl:
```diff
@@ -0,0 +1,82 @@
+	 
+	 float4 hueMinMax = float4(0.0f, 0.0f, 0.0f, 0.04f);//x = hueMin_1, y = hueMin_2, z = hueMax_1, w = hueMax_2;
+	 
+	 if ( ColorFilterValues.z > ColorFilterValues.y )
```
are we branching based on constant params across the entire frame? Because if we define them as uniform, I think the branching has no impact

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/AllInOne.fx.hlsl:
```diff
@@ -0,0 +1,83 @@
+	float4 color = tex2D(TESR_RenderedBuffer, IN.UVCoord);
+	float4 Adaptation = tex2D(TESR_AvgLumaBuffer, 0.5);
+	//float GrayAdaptation = max(max(Adaptation.x, Adaptation.y), 0.01);
+	float GrayAdaptation = max(tex2D(TESR_AvgLumaBuffer, float2(0.5, 0.5)).g, 0.01);
```
we're sampling the same buffer twice, we could probably save the sample into a variable instead

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/AllInOne.fx.hlsl:
```diff
@@ -0,0 +1,83 @@
+	//float GrayAdaptation = max(max(Adaptation.x, Adaptation.y), 0.01);
+	float GrayAdaptation = max(tex2D(TESR_AvgLumaBuffer, float2(0.5, 0.5)).g, 0.01);
+
+	float greyscale = dot(color.xyz, float3(0.3, 0.59, 0.11));
```
I think this should use the luma formula from Includes/Helpers.hlsl

mchaptel commented on Extra/Shaders/NewVegasReloaded/Effects/AllInOne.fx.hlsl:
```diff
@@ -0,0 +1,83 @@
+
+	float greyscale = dot(color.xyz, float3(0.3, 0.59, 0.11));
+
+	ApplyDesaturation(color.xyz, greyscale, TESR_AllInOneDesat.rgb);
```
does hlsl pass floats per reference and not per value?

It must other wise these functions don't do anything. Weird though
